### PR TITLE
chore(observability): robustly synchronize component validation framework tasks

### PR DESCRIPTION
**Pull request by neuronull**
TLDR:
- A separate task shutdown order was introduced to handle sinks, which require knowing the external resource processed all the events.
- The sink error path was changed to not rely on codecs but to rely on the external resource to emulate an error.

This change does a couple things that were side effects of the main goal, which was to remove the sleep that was added in the runner for the case of sinks. That was a temporary hack to get us going but as was seen in https://github.com/vectordotdev/vector/pull/19887 , it's not a long term solution as we needed to increase the sleep there due to the sink taking longer to process things.

In [this discussion](https://github.com/vectordotdev/vector/pull/17980#discussion_r1474617462) , we talked about synchronizing the sources but not the sinks.

The first attempt I made was in the `TaskCoordinator` , to split `shutdown()` into two methods- one to trigger the shutdown and the other to wait for it to complete. After that I played around with various orders of things, triggering the input task to shutdown, triggering topology to shutdown, etc. None of these fixed the issues we were having, in most cases they made other problems arise.

The "root cause" of the problem for sinks is that we need to have time for the sink to process all the events, the telemetry events to be received, AND for the external resource to process all the events. The only way to do this that I could see, was to drive the shutdown for sinks based on collecting the output task's result. Once we have that, we know the external resource has handled everything. Then it is safe to shutdown the component under test and the other tasks. As such, the solution was to have two different shutdown orders for sources/transforms, and for sinks.

The side effect of this change, was that it became obvious to me that the current solution for triggering errors in sinks is not ideal and not as representative of real world scenarios. This was discussed a bit in https://github.com/vectordotdev/vector/pull/18062. The current approach also raised challenges for the error path case in synchronizing the tasks shutdown, because we needed a way to know that the error had been raised by the sink somehow. The solution to this was to not leverage a codec error for the sinks, but instead an error from the external resource. This involved passing down the test case data to the external resource to be able to know when it needs to simulate an error.

```
TEST_LOG=info cargo nextest run --fail-fast --retries=0 --no-default-features --features component-validation-tests --status-level pass -j 4  components::validation::tests

    Starting 3 tests across 3 binaries (321 skipped)
        PASS [   5.778s] vector components::validation::tests::validate_component_tests_validation_components_sources_http_server_yaml
        PASS [  10.459s] vector components::validation::tests::validate_component_tests_validation_components_sources_http_client_yaml
        PASS [  11.709s] vector components::validation::tests::validate_component_tests_validation_components_sinks_http_yaml
------------
```

### CODE REVIEW

File: src/sinks/http/config.rs [hunk 1]
```diff
@@ -312,79 +312,39 @@ impl ValidatableComponent for HttpSinkConfig {
...
-            vec![
-                ComponentTestCaseConfig::from_sink(
-                    happy_config,
-                    None,
-                    Some(happy_external_resource),
-                ),
-                // this config only runs with the test case "encoding_error" in the yaml file.
-                ComponentTestCaseConfig::from_sink(
-                    sad_config,
-                    Some("encoding_error".to_owned()),
```
Comment (neuronull) [inline src/sinks/http/config.rs:384]: Note that with this PR, there is currently nothing using the functionality I added to tie a specific config to a specific test case(s). But, I think this could be valuable in the future so figured I'd leave the plumbing for that in place.

File: src/components/validation/sync.rs [hunk 2]
```diff
@@ -233,27 +235,31 @@ impl TaskCoordinator<Configuring> {
...
+            name: self.name,
         }
     }
 }
 
 impl TaskCoordinator<Started> {
     /// Triggers all coordinated tasks to shutdown, and waits for them to mark themselves as completed.
-    pub async fn shutdown(mut self) {
+    pub async fn shutdown(&mut self) {
+        info!("Triggering {} task to shutdown.", self.name);
```
Comment (tobz) [inline src/components/validation/sync.rs:246]: FWIW, you might consider a logging format of `{name}: <message here>` so that the task name is always the prefix. Easier for filtering/grepping logs.

File: src/components/validation/runner/telemetry.rs [hunk 3]
```diff
@@ -145,7 +145,6 @@ impl Telemetry {
             }
 
             grpc_task_coordinator.shutdown().await;
-            debug!("GRPC task(s) have been shutdown.");
```
Comment (tobz) [inline src/components/validation/runner/telemetry.rs:148]: This is perhaps a nit, but: why remove this?

By calling `shutdown`, you _are_ ensuring that the tasks that registered themselves with this coordinator -- so long as they're marking themselves as started and complete at the right points -- have in fact shutdown, since it does both the shutdown trigger and waits for all children to complete.

File: src/components/validation/runner/mod.rs [hunk 4]
```diff
@@ -400,6 +388,53 @@ impl Runner {
...
+
+            topology_task_coordinator.shutdown().await;
+
+            output_task_coordinator.shutdown().await;
+
+            output_driver
+                .await
+                .expect("output driver task should not have panicked")
+        }
+    }
```
Comment (tobz) [inline src/components/validation/runner/mod.rs:437]: I might be misreading this, but I wonder: could you not just use the first approach for all cases?

Like, if the first approach is working for the purpose of collecting/waiting long enough for all output events as well as telemetry data... then I'm trying to think of why it also wouldn't work for sources and transforms? 🤔

File: src/components/validation/resources/http.rs [hunk 5]
```diff
@@ -257,13 +267,26 @@ fn spawn_output_http_server(
...
+
     let (_, http_server_shutdown_tx) = spawn_http_server(
         task_coordinator,
         &config,
         runner_metrics,
         move |request, output_runner_metrics| {
             let output_tx = output_tx.clone();
             let mut decoder = decoder.clone();
+            let mut decoded_events = 0;
+            let server_sent_all_tx = server_sent_all_tx.clone();
```
Comment (tobz) [inline src/components/validation/resources/http.rs:289]: For what it's worth, this is only going to work on a per-connection basis.

If you ever ended up with a sink that, for example, gets N events in (where you expected to receive all N of those events here) but sends them across M requests (M > N), then this logic wouldn't ever get triggered, I don't think?

Like it might work now based on how things are batched, but it _feels_ a little shaky. I think it might just be easier to perhaps send the actual count of events processed over the channel, and tally those counts up down near `if expected_output_events > 0 {`... then you'd stop trying to receive once you've hit the intended received event count.

File: src/components/validation/resources/http.rs [hunk 6]
```diff
@@ -312,14 +353,25 @@ fn spawn_output_http_server(
...
+        info!("HTTP server external output resource started.");
 
+        // Wait for the runner to tell us to shutdown
         resource_shutdown_rx.wait().await;
-        _ = http_server_shutdown_tx.send(());
+
+        // ensure we've processed all events we expect to receive from the sink, if any.
+        if expected_output_events > 0 {
+            server_sent_all_rx.recv().await;
+        }
```
Comment (tobz) [inline src/components/validation/resources/http.rs:364]: Separate from my comment above about restructuring this somewhat to make the logic more durable... one thing that is _perhaps_ worth keeping in mind (but is just a comment here, not a thing to fix/address) is that this _does_ break the ability for the validation runner to gracefully kill tests, since we'll be stuck waiting here until N events are processed before we ever instruct the HTTP server to actually stop.

File: src/components/validation/resources/http.rs [hunk 7]
```diff
@@ -257,13 +267,26 @@ fn spawn_output_http_server(
...
     let decoder = codec.into_decoder()?;
 
+    // the number of events we expect to receive from the sink.
+    // any happy path events + any events that are flagged as external resource should reject.
+    let expected_output_events = input_events
+        .iter()
+        .filter(|te| !te.should_fail() || te.should_reject())
+        .count();
+
+    let should_reject = input_events.iter().filter(|te| te.should_reject()).count() > 0;
```
Comment (tobz) [inline src/components/validation/resources/http.rs:277]: One thing that jumps out to me here is that the logic, even though it may appear to, doesn't actually allow for differentiation between good and bad events. All events would be rejected even if only one of the input events is actually marked to cause a rejection.

This is another comment that isn't meant to indicate that I think you need to fix/change this, but just something worth thinking about.

**Review by tobz — APPROVED** [review]
Overall, I think this is generally fine.

I left some comments around areas that perhaps felt a little too specialized to the test cases you were intending to fix, because it feels like we might be trading one bit of complexity for another in some cases... but none of it is blocking, and is really just meant to point out where to be mindful in the future during any further refactoring.

File: src/components/validation/runner/telemetry.rs (hunk 3 above)
Comment (neuronull) [reply src/components/validation/runner/telemetry.rs:148]: Ah, I removed this because I think it's redundant with the changes I made to `TaskCoordinator` in this PR- in the shutdown() method (called the line above), we log an `info` level message that says the task has been shutdown.

File: src/components/validation/resources/http.rs (hunk 5 above)
Comment (neuronull) [reply src/components/validation/resources/http.rs:289]: Yes indeed. I did run into this on my branch based from this branch for the splunk_hec logs sink.

It has a couple other changes in the framework (including the one you identified here). So I'll probably need to split that out for a review as well. I'll address it in that PR since I already have changes in that branch.

File: src/components/validation/resources/http.rs (hunk 6 above)
Comment (neuronull) [reply src/components/validation/resources/http.rs:364]: Mmm I see what you mean yeah. We'd not get the output from the runner to know what metrics we did capture and where the discrepancies were. I think this could be solved by a timeout mechanism, I'll tack on on to the next PR since this one is approved.

File: src/components/validation/resources/http.rs (hunk 7 above)
Comment (neuronull) [reply src/components/validation/resources/http.rs:277]: Right, this did cross my mind as well. It is worth a code comment, I'll add one. The reason I didn't go down that path was because the test cases so far currently don't require that functionality, but we certainly might need it. And the way to solve it would be a bit tricky because we can't guarantee the order of events received (I believe). So we'd have to have some kind of tagging mechanism in the test case events that trickles down into here to say "if you receive an event with this field, reject it".

File: src/components/validation/runner/mod.rs (hunk 4 above)
Comment (neuronull) [reply src/components/validation/runner/mod.rs]: I had this thought as well when I was writing this. I had tried what you suggested and it didn't work for sources. I just tried it again and I realized there is a 3 birds one stone solution to the issues you have raised.

Instead of having the external resource be responsible for reporting if events received equals expected, we can do that in the output driver itself. This means sources will be able to use the same shutdown order, because the controlled edge doesn't need to coordinate things. And it means any other external resource implementation doesn't need to have that logic repeated.

It also presented an opportunity to add a timeout mechanism so that we still gracefully shutdown.

File: src/components/validation/resources/http.rs (hunk 6 above)
Comment (neuronull) [reply src/components/validation/resources/http.rs]: resolved in latest changes, we now have a timeout

File: src/components/validation/resources/http.rs (hunk 5 above)
Comment (neuronull) [reply src/components/validation/resources/http.rs]: This is resolved now that the output driver itself is tracking the event expectations
